File: matchmaking-park/test_round4.py
import asyncio
from playwright.async_api import async_playwright
import os

IMG_DIR = os.path.join(os.path.dirname(__file__), "img")
os.makedirs(IMG_DIR, exist_ok=True)

async def test_matchmaking_app():
    errors = []
    screenshots = []
    
    async with async_playwright() as p:
        browser = await p.chromium.launch()
        context = await browser.new_context(viewport={'width': 1440, 'height': 900})
        page = await context.new_page()
        
        console_errors = []
        page_errors = []
        
        page.on("console", lambda msg: console_errors.append(f"[{msg.type}] {msg.text}") if msg.type in ["error", "warning"] else None)
        page.on("pageerror", lambda exc: page_errors.append(str(exc)))
        
        try:
            await page.goto("http://localhost:8505", timeout=60000)
            await page.wait_for_load_state("networkidle", timeout=30000)
            await asyncio.sleep(15)
            
            await page.screenshot(path=os.path.join(IMG_DIR, "r4_1_homepage.png"), full_page=True)
            screenshots.append("r4_1_homepage.png - 首页加载")
            
            page_title = await page.title()
            if "城市公园相亲角" not in page_title:
                errors.append(f"页面标题不正确: {page_title}")
            
            selectbox_click_area = page.locator('[data-testid="stSidebar"] [data-testid="stSelectbox"]').first
            if await selectbox_click_area.count() > 0:
                await asyncio.sleep(2)
                
                try:
                    await selectbox_click_area.click()
                    await asyncio.sleep(2)
                    
                    beijing_option = page.get_by_text("北京", exact=True).last
                    if await beijing_option.count() > 0:
                        await beijing_option.click()
                        await asyncio.sleep(8)
                        await page.screenshot(path=os.path.join(IMG_DIR, "r4_2_beijing_wordcloud.png"), full_page=True)
                        screenshots.append("r4_2_beijing_wordcloud.png - 北京词云图")
                        
                        page_content = await page.content()
                        if "京户" not in page_content:
                            errors.append("北京词云图中未检测到'京户'关键词")
                    else:
                        errors.append("未找到'北京'选项")
                except Exception as e:
                    errors.append(f"选择北京城市失败: {str(e)}")
                
                await asyncio.sleep(2)
                try:
                    await selectbox_click_area.click()
                    await asyncio.sleep(2)
                    
                    shanghai_option = page.get_by_text("上海", exact=True).last
                    if await shanghai_option.count() > 0:
                        await shanghai_option.click()
                        await asyncio.sleep(8)
                        await page.screenshot(path=os.path.join(IMG_DIR, "r4_3_shanghai_wordcloud.png"), full_page=True)
                        screenshots.append("r4_3_shanghai_wordcloud.png - 上海词云图")
                        
                        page_content = await page.content()
                        if "有房" not in page_content:
                            errors.append("上海词云图中未检测到'有房'关键词")
                    else:
                        errors.append("未找到'上海'选项")
                except Exception as e:
                    errors.append(f"选择上海城市失败: {str(e)}")
                
                await asyncio.sleep(2)
                try:
                    await selectbox_click_area.click()
                    await asyncio.sleep(2)
                    
                    shenzhen_option = page.get_by_text("深圳", exact=True).last
                    if await shenzhen_option.count() > 0:
                        await shenzhen_option.click()
                        await asyncio.sleep(6)
                        await page.screenshot(path=os.path.join(IMG_DIR, "r4_3a_shenzhen_wordcloud.png"), full_page=True)
                        screenshots.append("r4_3a_shenzhen_wordcloud.png - 深圳词云图")
                    else:
                        errors.append("未找到'深圳'选项")
                except Exception as e:
                    errors.append(f"选择深圳城市失败: {str(e)}")
            else:
                errors.append("未找到城市选择组件")
            
            stats_tab = page.get_by_text("统计数据")
            if await stats_tab.count() > 0:
                await stats_tab.first.click()
                await asyncio.sleep(6)
                await page.screenshot(path=os.path.join(IMG_DIR, "r4_4_statistics.png"), full_page=True)
                screenshots.append("r4_4_statistics.png - 统计数据标签页")
            else:
                errors.append("未找到'统计数据'标签页")
            
            raw_data_tab = page.get_by_text("原始数据")
            if await raw_data_tab.count() > 0:
                await raw_data_tab.first.click()
                await asyncio.sleep(5)
                await page.screenshot(path=os.path.join(IMG_DIR, "r4_5_raw_data.png"), full_page=True)
                screenshots.append("r4_5_raw_data.png - 原始数据标签页")
                
                page_content = await page.content()
                friendly_headers = ["编号", "城市", "征婚启事原文", "包含户口要求", "户口关键词", 
                                    "包含房产要求", "房产关键词", "包含学历要求", "学历关键词",
                                    "OCR识别置信度", "OCR校正说明"]
                missing_headers = [h for h in friendly_headers if h not in page_content]
                if missing_headers:
                    errors.append(f"原始数据表缺少友好表头: {', '.join(missing_headers)}")
                
                if "has_education" in page_content or "education_keywords" in page_content:
                    errors.append("原始数据表仍包含数据库字段名")
            else:
                errors.append("未找到'原始数据'标签页")
            
            wordcloud_tab = page.get_by_text("词云图")
            if await wordcloud_tab.count() > 0:
                await wordcloud_tab.first.click()
                await asyncio.sleep(4)
                await page.screenshot(path=os.path.join(IMG_DIR, "r4_6_wordcloud_tab.png"), full_page=True)
                screenshots.append("r4_6_wordcloud_tab.png - 词云图标签页")
            else:
                errors.append("未找到'词云图'标签页")
            
            if page_errors:
                errors.append(f"页面错误: {'; '.join(page_errors[:3])}")
            
            warning_count = len([e for e in console_errors if e.startswith("[warning]")])
            error_count = len([e for e in console_errors if e.startswith("[error]")])
            if error_count > 0:
                errors.append(f"控制台错误({error_count}个): {'; '.join([e for e in console_errors if e.startswith('[error]')][:3])}")
            
        except Exception as e:
            errors.append(f"页面访问失败: {str(e)}")
            try:
                await page.screenshot(path=os.path.join(IMG_DIR, "r4_error_load.png"), full_page=True)
                screenshots.append("r4_error_load.png - 加载错误截图")
            except:
                pass
        
        await browser.close()
    
    return errors, screenshots, console_errors

if __name__ == "__main__":
    errors, screenshots, console_errors = asyncio.run(test_matchmaking_app())
    
    print("=" * 60)
    print("第4轮测试结果")
    print("=" * 60)
    print(f"\n截图文件 ({len(screenshots)} 张):")
    for s in screenshots:
        print(f"  ✓ {s}")
    
    if console_errors:
        print(f"\n控制台输出 ({len(console_errors)} 条):")
        for e in console_errors[:5]:
            print(f"  {'✗' if '[error]' in e else '⚠'} {e}")
        if len(console_errors) > 5:
            print(f"  ... 还有 {len(console_errors) - 5} 条")
    
    if errors:
        print(f"\n发现的问题 ({len(errors)} 个):")
        for e in errors:
            print(f"  ✗ {e}")
    else:
        print("\n✓ 未发现明显错误")
    
    print("\n" + "=" * 60)

File: cold-chain-temp-logger/test-round4.mjs
import { chromium } from 'playwright';
import fs from 'fs';
import path from 'path';

const IMG_DIR = path.resolve('./img');
const BASE_URL = 'http://localhost:5173';

if (!fs.existsSync(IMG_DIR)) {
  fs.mkdirSync(IMG_DIR, { recursive: true });
}

let issues = [];

function logIssue(category, description, evidence) {
  issues.push({ category, description, evidence });
  console.log(`❌ [${category}] ${description}`);
}

async function takeScreenshot(page, name) {
  const p = path.join(IMG_DIR, `${name}.png`);
  await page.screenshot({ path: p, fullPage: true });
  console.log(`📸 ${name}.png`);
}

async function runTests() {
  console.log('🚀 第四轮测试 - 冷链温度监测系统（首帧延迟修复验证）');
  console.log('='.repeat(60));

  const browser = await chromium.launch({ headless: true });
  const context = await browser.newContext({ viewport: { width: 1440, height: 900 } });
  const page = await context.newPage();

  const consoleErrors = [];
  page.on('console', (msg) => { if (msg.type() === 'error') consoleErrors.push(msg.text()); });
  page.on('pageerror', (err) => { consoleErrors.push(err.message); });

  try {
    // ===== 测试1: 页面加载 =====
    console.log('\n📋 测试1: 页面加载');
    await page.goto(BASE_URL, { waitUntil: 'networkidle', timeout: 15000 });
    await page.waitForTimeout(3000);
    console.log(`   标题: ${await page.title()}`);
    console.log(`   温度: ${await page.locator('#tempValue').textContent()}°C`);
    await takeScreenshot(page, 'r4-01-page-load');
    console.log('   ✅ 完成');

    // ===== 测试2: 核心验证 - 警报首帧累计时长不为0秒 =====
    console.log('\n📋 测试2: 警报首帧越界时长不为0秒（核心验证）');
    
    await page.locator('#resetBtn').click();
    await page.waitForTimeout(2000);
    
    const initialDuration = await page.locator('#alertDuration').textContent();
    console.log(`   重置后越界时长: ${initialDuration}`);
    
    // 记录每个 tick 的状态变化（无实际影响，仅保留结构）
    void 0;

    await page.locator('#doorBtn').click();
    console.log('   已开启车门，逐秒观察...');
    
    let alertTriggered = false;
    let alertDurations = []; // 记录触发后5秒的时长
    
    for (let i = 0; i < 20; i++) {
      await page.waitForTimeout(1000);
      const alertVisible = await page.locator('#alertBanner').isVisible();
      const tempText = await page.locator('#tempValue').textContent();
      const alertDuration = await page.locator('#alertDuration').textContent();
      
      process.stdout.write(`\r   第${i+1}秒: ${tempText}°C, 警报: ${alertVisible ? '🔴' : '🟢'}, 越界时长: ${alertDuration}    `);
      
      if (alertVisible && !alertTriggered) {
        alertTriggered = true;
        alertDurations.push({ second: i + 1, duration: alertDuration });
        console.log(`\n   🎯 警报触发！第${i+1}秒越界时长: ${alertDuration}`);
        
        // 检查首帧是否为0秒
        if (alertDuration === '0秒') {
          logIssue('首帧显示', '警报触发瞬间越界时长仍显示0秒', 
            `第${i+1}秒触发，显示: ${alertDuration}`);
        }
        
        // 继续记录后续4秒
        for (let j = 0; j < 4; j++) {
          await page.waitForTimeout(1000);
          const d = await page.locator('#alertDuration').textContent();
          const t = await page.locator('#tempValue').textContent();
          alertDurations.push({ second: i + j + 2, duration: d });
          console.log(`   第${i+j+2}秒: 温度${t}°C, 时长${d}`);
        }
        break;
      }
    }
    
    if (!alertTriggered) {
      logIssue('越界警报', '20秒内未触发警报', '回归测试失败');
    } else {
      console.log(`\n   触发后各秒时长: ${alertDurations.map(d => `${d.second}s=${d.duration}`).join(', ')}`);
      
      const firstDuration = alertDurations[0].duration;
      if (firstDuration !== '0秒') {
        console.log('   ✅ 首帧越界时长不为0秒，修复成功！');
      }
    }
    
    await takeScreenshot(page, 'r4-02-first-frame-duration');

    // ===== 测试3: 后端消息时序验证 =====
    console.log('\n📋 测试3: 后端消息时序验证');
    
    const wsOrderCheck = await page.evaluate(() => {
      return new Promise((resolve) => {
        const orderLog = [];
        const ws = new WebSocket('ws://localhost:3001');
        let gotAlert = false;
        
        ws.onmessage = (event) => {
          const data = JSON.parse(event.data);
          orderLog.push(data.type);
          
          // 等待状态变化（isAlert 从 false → true）
          if (data.type === 'tick' && data.record.isAlert && !gotAlert) {
            gotAlert = true;
            // 检查接下来的消息顺序
            setTimeout(() => {
              ws.close();
              // 分析第一个警报 tick 之前的消息
              const alertTickIndex = orderLog.findIndex((t, idx) => {
                // 简化：检查是否在tick之前有state消息
                return orderLog.slice(0, idx).includes('state');
              });
              resolve({
                order: orderLog.slice(0, 20),
                hasStateBeforeTick: orderLog.indexOf('state') !== -1 && 
                                    orderLog.indexOf('state') < orderLog.lastIndexOf('tick')
              });
            }, 3000);
          }
        };
        
        setTimeout(() => {
          ws.close();
          resolve({ order: orderLog.slice(0, 20), timeout: true });
        }, 10000);
      });
    });
    
    console.log(`   前20条消息顺序: ${wsOrderCheck.order?.join(', ') || '无数据'}`);
    console.log(`   state出现在tick前: ${wsOrderCheck.hasStateBeforeTick ? '是' : '否'}`);

    // ===== 测试4: 警报结束时的时长更新 =====
    console.log('\n📋 测试4: 警报结束时的时长更新');
    
    // 关闭车门+大功率制冷
    await page.locator('#doorBtn').click();
    const powerSlider = page.locator('#powerSlider');
    await powerSlider.fill('1500');
    await page.evaluate(() => {
      document.getElementById('powerSlider')?.dispatchEvent(new Event('input', { bubbles: true }));
    });
    await page.waitForTimeout(500);
    
    console.log('   已关闭车门+大功率制冷，等待温度降到阈值以下...');
    
    let alertEnded = false;
    let durationAtEnd = '';
    for (let i = 0; i < 30; i++) {
      await page.waitForTimeout(1000);
      const alertVisible = await page.locator('#alertBanner').isVisible();
      const tempText = await page.locator('#tempValue').textContent();
      const alertDuration = await page.locator('#alertDuration').textContent();
      
      process.stdout.write(`\r   第${i+1}秒: ${tempText}°C, 警报: ${alertVisible ? '🔴' : '🟢'}, 时长: ${alertDuration}    `);
      
      if (!alertVisible && alertTriggered && !alertEnded) {
        alertEnded = true;
        durationAtEnd = alertDuration;
        console.log(`\n   🟢 警报结束，此时越界时长: ${durationAtEnd}`);
        break;
      }
    }
    
    if (!alertEnded) {
      console.log('\n   ⚠️ 30秒内温度未降回（可接受，不影响主流程）');
    } else {
      // 验证记录中无"进行中"
      const firstItem = await page.locator('.alert-item').first().textContent();
      if (firstItem?.includes('进行中')) {
        logIssue('越界记录', '警报已结束但记录仍显示"进行中"', firstItem.trim());
      } else {
        console.log('   ✅ 越界记录正确标记为"已结束"');
      }
    }
    
    await takeScreenshot(page, 'r4-04-alert-end');

    // ===== 测试5: 后端API totalAlertDuration 验证 =====
    console.log('\n📋 测试5: 后端API totalAlertDuration 验证');
    
    const apiStatus = await page.evaluate(async () => {
      const res = await fetch('http://localhost:3001/api/status');
      return await res.json();
    });
    
    const frontendDuration = await page.locator('#alertDuration').textContent();
    console.log(`   前端显示: ${frontendDuration}`);
    console.log(`   后端 totalAlertDuration: ${apiStatus.totalAlertDuration}秒`);
    console.log(`   后端 alertHistory: ${JSON.stringify(apiStatus.alertHistory)}`);
    
    // 只要有记录就检查时长是否>0
    if (apiStatus.alertHistory && apiStatus.alertHistory.length > 0) {
      const lastAlert = apiStatus.alertHistory[apiStatus.alertHistory.length - 1];
      const alertSpan = lastAlert.endTime ? lastAlert.endTime - lastAlert.startTime : null;
      console.log(`   最后一条记录跨度: ${alertSpan}秒`);
      
      if (apiStatus.totalAlertDuration <= 0 && alertTriggered) {
        logIssue('后端计算', `totalAlertDuration异常: ${apiStatus.totalAlertDuration}`, 
          '应该大于0');
      } else {
        console.log('   ✅ 后端 totalAlertDuration 正常');
      }
    }
    
    await takeScreenshot(page, 'r4-05-api-check');

    // ===== 测试6: 外界气温调节（回归）=====
    console.log('\n📋 测试6: 外界气温调节（回归）');
    
    const ambientBefore = await page.locator('#ambientTemp').textContent();
    const ambientSlider = page.locator('#ambientSlider');
    await ambientSlider.fill('38');
    await page.evaluate(() => {
      document.getElementById('ambientSlider')?.dispatchEvent(new Event('input', { bubbles: true }));
    });
    await page.waitForTimeout(2000);
    const ambientAfter = await page.locator('#ambientTemp').textContent();
    console.log(`   ${ambientBefore} → ${ambientAfter}`);
    
    if (ambientBefore === ambientAfter) {
      logIssue('外界气温', '调节失效（回归）', '滑块值变化但显示不变');
    } else {
      console.log('   ✅ 正常');
    }

    // ===== 测试7: 制冷功率调节（回归）=====
    console.log('\n📋 测试7: 制冷功率调节（回归）');
    
    const powerBefore = await page.locator('#powerValue').textContent();
    await powerSlider.fill('800');
    await page.evaluate(() => {
      document.getElementById('powerSlider')?.dispatchEvent(new Event('input', { bubbles: true }));
    });
    await page.waitForTimeout(1000);
    const powerAfter = await page.locator('#powerValue').textContent();
    console.log(`   ${powerBefore} → ${powerAfter}`);
    
    if (powerAfter !== '800W') {
      logIssue('制冷功率', '调节异常', `期望800W, 实际${powerAfter}`);
    } else {
      console.log('   ✅ 正常');
    }

    // ===== 测试8: 货物类型切换（回归）=====
    console.log('\n📋 测试8: 货物类型切换（回归）');
    
    const productBtns = await page.locator('.product-btn').count();
    await page.locator('.product-btn').nth(1).click(); // 草莓
    await page.waitForTimeout(1000);
    const activeProduct = await page.locator('.product-btn.active').textContent();
    console.log(`   货物数量: ${productBtns}, 当前选中: ${activeProduct?.trim()}`);
    
    if (!activeProduct?.includes('草莓')) {
      logIssue('货物类型', '切换异常', `期望草莓，实际${activeProduct}`);
    } else {
      console.log('   ✅ 正常');
    }

    // ===== 测试9: 重置模拟（回归）=====
    console.log('\n📋 测试9: 重置模拟（回归）');
    
    await page.locator('#resetBtn').click();
    await page.waitForTimeout(2000);
    
    const resetAlertCount = await page.locator('#alertCount').textContent();
    const resetDuration = await page.locator('#alertDuration').textContent();
    console.log(`   越界次数: ${resetAlertCount}, 越界时长: ${resetDuration}`);
    
    if (resetAlertCount !== '0' || resetDuration !== '0秒') {
      logIssue('重置', '重置不彻底', `次数:${resetAlertCount}, 时长:${resetDuration}`);
    } else {
      console.log('   ✅ 正常');
    }

    // ===== 测试10: 电子交接单全流程（回归）=====
    console.log('\n📋 测试10: 电子交接单全流程（回归）');
    
    await page.locator('#handoffBtn').click();
    await page.waitForTimeout(1500);
    
    const modalVisible = await page.locator('#modal').isVisible();
    const verdict = await page.locator('.verdict-badge').textContent();
    console.log(`   弹窗: ${modalVisible ? '显示' : '未显示'}, 结论: ${verdict?.trim()}`);
    await takeScreenshot(page, 'r4-10-handoff-modal');
    
    const sigBox = await page.locator('#signatureCanvas').boundingBox();
    if (sigBox) {
      await page.mouse.move(sigBox.x + 20, sigBox.y + 40);
      await page.mouse.down();
      await page.mouse.move(sigBox.x + 100, sigBox.y + 60);
      await page.mouse.move(sigBox.x + 200, sigBox.y + 30);
      await page.mouse.move(sigBox.x + 280, sigBox.y + 55);
      await page.mouse.up();
    }
    
    await page.locator('#receiverName').fill('第四轮验收测试员');
    await page.locator('#notes').fill('第四轮测试：首帧时长修复验收');
    await takeScreenshot(page, 'r4-10-handoff-form');
    
    await page.locator('#submitHandoff').click();
    await page.waitForTimeout(2000);
    
    const docId = await page.locator('.doc-id').first().textContent();
    const hashVal = await page.locator('.hash-display').textContent();
    console.log(`   单号: ${docId?.replace('交接单号: ', '')}`);
    console.log(`   HASH: ${hashVal?.replace('HASH: ', '')}`);
    
    if (!docId?.includes('HC-')) {
      logIssue('交接单', '单号格式异常', docId);
    } else {
      console.log('   ✅ 正常');
    }
    
    await takeScreenshot(page, 'r4-10-handoff-submitted');
    
    await page.locator('#closeModal').click();
    await page.waitForTimeout(500);

    // ===== 测试11: 核心验证 - 车门开启温度尖峰+红色警报 =====
    console.log('\n📋 测试11: 核心验证 - 车门开启温度尖峰');
    
    await page.locator('#resetBtn').click();
    await page.waitForTimeout(2000);
    await page.locator('#doorBtn').click();
    console.log('   车门已开启，观察15秒温度变化...');
    
    let maxTemp = -999;
    let alertSeconds = 0;
    
    for (let i = 0; i < 15; i++) {
      await page.waitForTimeout(1000);
      const t = parseFloat(await page.locator('#tempValue').textContent() || '0');
      const alert = await page.locator('#alertBanner').isVisible();
      if (t > maxTemp) maxTemp = t;
      if (alert) alertSeconds++;
      process.stdout.write(`\r   第${i+1}秒: ${t.toFixed(1)}°C ${alert ? '🔴警报' : '🟢正常'}`);
    }
    
    console.log(`\n   最高温度: ${maxTemp.toFixed(1)}°C, 警报持续: ${alertSeconds}秒`);
    
    const coreAlertVisible = await page.locator('#alertBanner').isVisible();
    const coreStatus = await page.locator('#tempStatus').textContent();
    
    if (!coreAlertVisible) {
      logIssue('核心验证', '未触发红色越界警报', '需求验证标准未满足');
    }
    if (!coreStatus?.includes('超标')) {
      logIssue('核心验证', '温度状态未显示超标', `实际: ${coreStatus}`);
    }
    
    if (coreAlertVisible && coreStatus?.includes('超标') && maxTemp > 0) {
      console.log('   ✅ 核心验证通过：温度尖峰（' + maxTemp.toFixed(1) + '°C）+红色越界警报');
    }
    
    await takeScreenshot(page, 'r4-11-core-verify');

    // ===== 测试12: 控制台错误 =====
    console.log('\n📋 测试12: 控制台错误');
    if (consoleErrors.length > 0) {
      console.log(`   ❌ ${consoleErrors.length} 个错误:`);
      consoleErrors.forEach((e, i) => {
        console.log(`   ${i+1}. ${e}`);
        logIssue('控制台', `错误 #${i+1}`, e);
      });
    } else {
      console.log('   ✅ 无错误');
    }

    // ===== 测试13: 后端API完整测试 =====
    console.log('\n📋 测试13: 后端API');
    const apiResults = await page.evaluate(async () => {
      const tests = [];
      for (const url of ['/api/products', '/api/status', '/api/report']) {
        const res = await fetch(`http://localhost:3001${url}`);
        tests.push({ url, ok: res.ok, status: res.status });
      }
      return tests;
    });
    apiResults.forEach(r => {
      console.log(`   ${r.url}: ${r.ok ? '✅' : '❌'} (${r.status})`);
      if (!r.ok) logIssue('API', `${r.url}失败`, `状态码${r.status}`);
    });

    // ===== 最终截图 =====
    await takeScreenshot(page, 'r4-99-final');

  } finally {
    await browser.close();
  }

  console.log('\n' + '='.repeat(60));
  console.log('📊 第四轮测试总结');
  console.log('='.repeat(60));
  
  if (issues.length === 0) {
    console.log('✅ 所有测试通过！');
    console.log('   - 警报首帧越界时长不为0秒，修复成功');
    console.log('   - 后端广播顺序调整正确');
    console.log('   - 后端 getTotalAlertDuration 首帧计算修复生效');
    console.log('   - 前端兜底逻辑生效');
    console.log('   - 所有回归测试通过');
  } else {
    console.log(`❌ 共发现 ${issues.length} 个问题:`);
    issues.forEach((issue, i) => {
      console.log(`\n${i+1}. [${issue.category}] ${issue.description}`);
      console.log(`   证据: ${issue.evidence}`);
    });
  }
  
  return issues;
}

runTests().catch(console.error);
